mlrun/api/launcher.py

```diff
@@ -28,16 +109,143 @@ def save(runtime):
+                    and run.metadata.project
+                    == mlrun.pipeline_context.project.metadata.name
+                ):
+                    run.spec.output_path = (
+                        mlrun.pipeline_context.project.spec.artifact_path
+                        or mlrun.pipeline_context.workflow_artifact_path
+                    )
+
+            if not run.spec.output_path and runtime._get_db():
+                try:
+                    # not passing or loading the DB before the enrichment on purpose, because we want to enrich the
+                    # spec first as get_db() depends on it
+                    project = runtime._get_db().get_project(run.metadata.project)
+                    run.spec.output_path = project.spec.artifact_path
+                except mlrun.errors.MLRunNotFoundError:
+                    logger.warning(
+                        f"project {run.metadata.project} is not saved in DB yet, "
+                        f"enriching output path with default artifact path: {mlrun.mlconf.artifact_path}"
+                    )
+
+        if not run.spec.output_path:
+            run.spec.output_path = mlrun.mlconf.artifact_path
+
+        if run.spec.output_path:
+            run.spec.output_path = run.spec.output_path.replace("{{run.uid}}", meta.uid)
+            run.spec.output_path = mlrun.utils.helpers.fill_artifact_path_template(
+                run.spec.output_path, run.metadata.project
+            )
+
+        run.spec.notifications = run.spec.notifications or []
+        return run
 
     @staticmethod
-    def _validate_runtime(runtime):
-        pass
+    def _validate_output_path(runtime: BaseRuntime, run: RunObject):
+        # TODO: move is_local somewhere else
+        def is_local(url):
+            if not url:
+                return True
+            return "://" not in url
+
+        if is_local(run.spec.output_path):
+            message = ""
+            if not os.path.isabs(run.spec.output_path):
+                message = (
+                    "artifact/output path is not defined or is local and relative,"
+                    " artifacts will not be visible in the UI"
+                )
+                if mlrun.runtimes.RuntimeKinds.requires_absolute_artifacts_path(
+                    runtime.kind
+                ):
+                    raise mlrun.errors.MLRunPreconditionFailedError(
+                        "artifact path (`artifact_path`) must be absolute for remote tasks"
+                    )
+            elif (
+                hasattr(runtime.spec, "volume_mounts")
+                and not runtime.spec.volume_mounts
+            ):
+                message = (
+                    "artifact output path is local while no volume mount is specified. "
+                    "artifacts would not be visible via UI."
+                )
+            if message:
+                logger.warning(message, output_path=run.spec.output_path)
```

Comment:
the `message` assignment and log warning seem redundant when running in server context, as the user won't see it anyway and it will spam the log without any related data, I would suggest removing it.

Reply:
this is irrelevant because we moved `_enrich_run` to the base class. if we want to separate client and server enrich run it can be done later.